### PR TITLE
Fix jmap request to server is called when query string is empty in quick search web

### PR DESCRIPTION
**Pull request by dab246**
## Issue

-  Jmap request to server is called when query string is empty in quick search web

https://github.com/user-attachments/assets/935216a5-8e7b-4b1c-88c2-cf7c17e08eb7

## Root cause

- Using `Timer` to debounce, but it doesn't work properly.

## Solution

-Using [debounce_throttle](https://pub.dev/packages/debounce_throttle) to debounce

## Resolved

https://github.com/user-attachments/assets/410f808f-2fa2-4a81-b8be-faa4226228c9

### CODE REVIEW

**Review by tddang-linagora — DISMISSED** [review]

**Comment by hoangdat** [comment]
- `da` is requested and `dat` also. It is not optimized for search

**Comment by dab246** [comment]
> * [x]  `da` is requested and `dat` also. It is not optimized for searching


https://github.com/user-attachments/assets/7b500fe8-da85-4aef-b815-9d93b0d0ea21


> * [x]  write the integration test to make sure: request only be sent after user stop typing for a while (Ex: 300 ms)
 
Currently, we don't have integration test on web.

File: core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+    }
+  }
+
+  Future<void> _handleDebounceTimeListener(String queryString) async {
+    log('QuickSearchSuggestionListState::_handleDebounceTimeListener:queryString = $queryString | minCharsForSuggestions = ${widget.minCharsForSuggestions}');
+    if (!mounted) return;
+
+    if (widget.minCharsForSuggestions != null &&
+        queryString.length <= widget.minCharsForSuggestions!) {
+      final recentItems = await _getListRecent(queryString);
```
Comment (tddang-linagora) [inline core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart:143]: Will this method cause race condition? If yes, should we implement cancel token? E.g. 
- Method 1 is `_handleDebounceTimeListener('abc')`
- Method 2 is `_handleDebounceTimeListener('abc123')`
- Method 1 is called before method 2, but method 1 response after method 2 response

Comment (dab246) [reply core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart:149]: Good catch. Fixed

**Review by tddang-linagora — DISMISSED** [review]

**Review by hoangdat — DISMISSED** [review]

File: core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+      return widget.fetchRecentActionCallback?.call(queryString);
+    } catch (e) {
+      logError('SuggestionsListState::_getRecent:Exception = $e');
+      return null;
+    }
+  }
+
+  Future<Iterable<P>?> _getListContact(String queryString) async {
+    try {
+      return widget.contactSuggestionCallback?.call(queryString);
```
Comment (hoangdat) [inline core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart:121]: we should `await` here?

File: core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+    } catch (e) {
+      logError('SuggestionsListState::_getListContact:Exception = $e');
+      return null;
+    }
+  }
+
+  Future<Iterable<T>?> _getListSuggestion(String queryString) async {
+    try {
+      return widget.suggestionsCallback?.call(queryString);
```
Comment (hoangdat) [inline core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart:130]: idem

File: core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+  AnimationController? _animationController;
+  String? _lastTextValue;
+  late final ScrollController _scrollController =
+      widget.scrollController ?? ScrollController();
+  late final Debouncer<String> _deBouncerSuggestion;
+  late final StreamSubscription<String> _deBouncerSuggestionStreamSubscriptions;
+
+  Future<Iterable<R>?> _getListRecent(String queryString) async {
+    try {
+      return widget.fetchRecentActionCallback?.call(queryString);
```
Comment (hoangdat) [inline core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart:112]: idem

File: core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart (hunk 2 above)
Comment (dab246) [reply core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart]: Yes, updated

File: core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart (hunk 3 above)
Comment (dab246) [reply core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart]: updated

File: core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart (hunk 4 above)
Comment (dab246) [reply core/lib/presentation/views/quick_search/quick_search_suggestion_list.dart]: updated

**Review by hoangdat — APPROVED** [review]

**Review by tddang-linagora — APPROVED** [review]

**Pull request merged**